File: decorator/java/Decorator.java
// Component interface
interface Coffee {
    String getDescription();
    double getCost();
}

// Concrete Component: Simple Coffee
class SimpleCoffee implements Coffee {
    @Override
    public String getDescription() {
        return "Simple Coffee";
    }

    @Override
    public double getCost() {
        return 2.0;
    }
}

// Decorator abstract class
abstract class CoffeeDecorator implements Coffee {
    protected Coffee coffee;

    public CoffeeDecorator(Coffee coffee) {
        this.coffee = coffee;
    }

    @Override
    public String getDescription() {
        return coffee.getDescription();
    }

    @Override
    public double getCost() {
        return coffee.getCost();
    }
}

// Concrete Decorator: Milk
class Milk extends CoffeeDecorator {
    public Milk(Coffee coffee) {
        super(coffee);
    }

    @Override
    public String getDescription() {
        return coffee.getDescription() + ", Milk";
    }

    @Override
    public double getCost() {
        return coffee.getCost() + 0.5;
    }
}

// Concrete Decorator: Sugar
class Sugar extends CoffeeDecorator {
    public Sugar(Coffee coffee) {
        super(coffee);
    }

    @Override
    public String getDescription() {
        return coffee.getDescription() + ", Sugar";
    }

    @Override
    public double getCost() {
        return coffee.getCost() + 0.2;
    }
}

// Main class demonstrating Decorator pattern
public class Decorator {
    public static void main(String[] args) {
        // Simple coffee
        Coffee coffee1 = new SimpleCoffee();
        System.out.println(coffee1.getDescription() + " - $" + coffee1.getCost());

        // Coffee with milk
        Coffee coffee2 = new Milk(new SimpleCoffee());
        System.out.println(coffee2.getDescription() + " - $" + coffee2.getCost());

        // Coffee with milk and sugar
        Coffee coffee3 = new Sugar(new Milk(new SimpleCoffee()));
        System.out.println(coffee3.getDescription() + " - $" + coffee3.getCost());

        // Coffee with double milk and sugar
        Coffee coffee4 = new Sugar(new Milk(new Milk(new SimpleCoffee())));
        System.out.println(coffee4.getDescription() + " - $" + coffee4.getCost());
    }
}

File: decorator/typescript/Decorator.ts
// Component interface
interface Coffee {
  getDescription(): string;
  getCost(): number;
}

// Concrete Component
class SimpleCoffee implements Coffee {
  getDescription(): string {
    return 'Simple Coffee';
  }

  getCost(): number {
    return 2.0;
  }
}

// Decorator
abstract class CoffeeDecorator implements Coffee {
  protected coffee: Coffee;

  constructor(coffee: Coffee) {
    this.coffee = coffee;
  }

  getDescription(): string {
    return this.coffee.getDescription();
  }

  getCost(): number {
    return this.coffee.getCost();
  }
}

// Concrete Decorators
class MilkDecorator extends CoffeeDecorator {
  getDescription(): string {
    return `${this.coffee.getDescription()}, Milk`;
  }

  getCost(): number {
    return this.coffee.getCost() + 0.5;
  }
}

class SugarDecorator extends CoffeeDecorator {
  getDescription(): string {
    return `${this.coffee.getDescription()}, Sugar`;
  }

  getCost(): number {
    return this.coffee.getCost() + 0.3;
  }
}

// Usage
let coffee: Coffee = new SimpleCoffee();
console.log(`${coffee.getDescription()}: $${coffee.getCost().toFixed(2)}`);

coffee = new MilkDecorator(coffee);
console.log(`${coffee.getDescription()}: $${coffee.getCost().toFixed(2)}`);

coffee = new SugarDecorator(coffee);
console.log(`${coffee.getDescription()}: $${coffee.getCost().toFixed(2)}`);
